### DISCUSSION TITLE
Correct approach to connect to Sharepoint Online from RestAPI(PnP Core SDK) published in azure.

### DISCUSSION THREAD

**Discussion by necik11** [pnp/pnpcore#1052] [Q&A]
Good day,

I would like an explanation and a correct approach to a specific example_

I have a plain RestAPI similar from the [tutorials ](https://learn.microsoft.com/en-us/aspnet/core/tutorials/first-web-api?view=aspnetcore-7.0&tabs=visual-studio)(placed in Azure).

In the methods, I use [PnP SDK Core](https://pnp.github.io/pnpcore/using-the-sdk/readme.html?q=PnPCoreSiteOptions) - to connect to Sharepoint Online and then we do complex logic, which may take 30m - calling other APIs + Opening various Sharepoint sites and setting items, permissions, etc...

Question:
1. **What is the correct login approach for multiple opening of the same context only on another site?  -**
a) Is it enough to use login, password, clientId, tenantId from appconfig?
b) Or use AppSecret in app config?
c) Paid function - Key Vault?
d) The worst/ most complicated variant Certificate?
e) Or use [persistent token](https://pnp.github.io/pnpcore/demos/Demo.PersistentTokenCache/README.html) - but there is just Interactive I need call in BackEnd...  Add to StartUp [AddInMemoryTokenCaches](https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-net-token-cache-serialization?tabs=aspnetcore)? Or [Acquire and cache tokens using the Microsoft Authentication Library (MSAL)?](https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-acquire-cache-tokens)

I have a lot of tutorials, articles, libraries but nowhere is it complete, always only a part of fragments that do not fit into the whole...

Is it possible advice?

**Comment by jansenbe** [comment]
@necik11 : if you've one context you can easily create a new context for anther site using the `Clone` methods, see https://pnp.github.io/pnpcore/using-the-sdk/basics-context.html#cloning-a-pnpcontext information on how to do that.

**Comment by necik11** [comment]
Yes but i am afraid of this:[ While threading is a very attractive option to speed up operations do keep in mind that your application might get throttled sooner as throttling protection kicks in whenever the amount of requests an application makes in a given time window exceeds a certain threshold. ](https://pnp.github.io/pnpcore/using-the-sdk/basics-threading.html?q=throttle).

I am not sure, i got such an error message.. when i opened the many CreateContext... So I think CloneContext doing similiary way and generate everytime new token... so thats why i am asking to another way ..
Microsoft.Identity.Client.MsalUiRequiredException: Your app has been throttled by AAD due to too many requests. To avoid this, cache your tokens see https://aka.ms/msal-net-throttling.

**Reply by jansenbe** [reply]
@necik11 : Please ensure you do initialize your function once (e.g. like shown here: https://github.com/pnp/pnpcore/blob/dev/samples/Demo.AzureFunction.OutOfProcess.AppOnly/Program.cs#L14-L73 or https://github.com/pnp/pnpcore/blob/dev/samples/Demo.AzFunction.ManagedIdentity/Startup.cs#L13-L67, in then uses the default MSAL in memory cache provider and will only acquire a new access token once the current one is about to expire.

**Reply by jansenbe** [reply]
@necik11 : also see this thread https://github.com/pnp/pnpcore/discussions/1024
